datafusion/common/src/scalar/mod.rs

```diff
@@ -2453,15 +2453,23 @@ impl ScalarValue {
                             value.to_array_of_size(size)?
                         } else {
                             let dt = field.data_type();
-                            new_null_array(dt, size)
+                            match mode {
+                                UnionMode::Sparse => new_null_array(dt, size),
+                                // In a dense union, only the child with values needs to be
```

Comment:
I double checked the spec and I agree with this check: https://arrow.apache.org/docs/format/Columnar.html#dense-union